metron-platform/Performance-tuning-guide.md

```diff
@@ -0,0 +1,326 @@
+enrichments                    enrichments        3          29754325        29754326        1               consumer-1_/xxx.xxx.xxx.xxx
+enrichments                    enrichments        43         29754331        29754332        1               consumer-6_/xxx.xxx.xxx.xxx
+...
+```
+
+_Note_: You won't see any output until a topology is actually running because the consumer groups only exist while consumers in the
+spouts are up and running.
+
+The primary column we're concerned with paying attention to is the LAG column, which is the current delta calculation between the
+current and end offset for the partition. This tells us how close we are to keeping up with incoming data. And, as we found through
+multiple trials, whether there are any problems with specific consumers getting stuck.
+
+Taking this one step further, it's probably more useful if we can watch the offsets and lags change over time. In order to do this
+we'll add a "watch" command and set the refresh rate to 10 seconds.
+
+```
+watch -n 10 -d ${KAFKA_HOME}/bin/kafka-consumer-groups.sh \
+    --command-config=/tmp/consumergroup.config \
+    --describe \
+    --group enrichments \
+    --bootstrap-server $BROKERLIST \
+    --new-consumer
+```
+
+Every 10 seconds the command will re-run and the screen will be refreshed with new information. The most useful bit is that the
+watch command will highlight the differences from the current output and the last output screens.
+
+We can also monitor our Storm topologies by using the Storm UI - see http://www.malinga.me/reading-and-understanding-the-storm-ui-storm-ui-explained/
+
+And lastly, you can leverage some GUI tooling to make creating and modifying your Kafka topics a bit easier -
+see https://github.com/yahoo/kafka-manager
+
+## General Knobs and Levers
+
+Kafka
+    - # partitions
+Storm
+    Kafka
+        - polling frequency and timeouts
+    - # workers
+    - ackers
+    - max spout pending
+    - spout parallelism
+    - bolt parallelism
+    - # executors
+Metron
+    - bolt cache size - handles how many messages can be cached. This cache is used while waiting for all parts of the message to be rejoined.
+
+## Topologies
+
+### Parsers
+
+The parsers and PCAP use a builder utility, as opposed to enrichments and indexing, which use Flux.
+
+We set the number of partitions for our inbound Kafka topics to 48.
+
+```
+$ cat ~metron/.storm/storm-bro.config
+
+{
+    ...
+    "topology.max.spout.pending" : 2000
+    ...
+}
+```
+
+These are the spout recommended defaults from Storm and are currently the defaults provided in the Kafka spout itself.
+In fact, if you find the recommended defaults work fine for you, then this file might not be necessary at all.
+```
+$ cat ~/.storm/spout-bro.config
+{
+    ...
+    "spout.pollTimeoutMs" : 200,
+    "spout.maxUncommittedOffsets" : 10000000,
+    "spout.offsetCommitPeriodMs" : 30000
+}
+```
+
+We ran our bro parser topology with the following options
```

Comment:
Is there any explanation of how we got to these numbers? There will be heavily dependent on hardware, particularly cpu and disk configuration, so the context could be very important.